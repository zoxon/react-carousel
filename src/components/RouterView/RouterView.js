import React from "react";
import PropTypes from "prop-types";

import history from "lib/history";
import { removeBasename } from "lib/utils";

const DEFAULT_PATHNAME = "*";

class RouterView extends React.Component {
  state = {
    pathname: "/"
  };

  componentDidMount() {
    this.setState({
      pathname: this.props.pathname
    });

    history.onChange(this.handleNav);
  }

  handleNav = pathname => {
    this.setState({ pathname });
  };

  render() {
    const path = removeBasename(
      this.props.basename,
      this.state.pathname
    ).replace(/\\\//gi, "/");

    const Component =
      this.props.routes[path] ||
      this.props.routes[DEFAULT_PATHNAME] ||
      (() => <span>Page not found</span>);

    return <Component context={this.state}>{this.props.children}</Component>;
  }
}

RouterView.defaultProps = {
  pathname: "/",
  basename: "/"
};

RouterView.propTypes = {
  routes: PropTypes.object.isRequired,
  pathname: PropTypes.string.isRequired,
  children: PropTypes.node,
  basename: PropTypes.string
};

export default RouterView;
