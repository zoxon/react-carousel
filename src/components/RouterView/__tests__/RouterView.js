import React from "react";
import { mount } from "enzyme";

import RouterView from "../RouterView";

const HomePage = props => <h1>Home page</h1>;
const NotFoundPage = props => <h1>Not found page</h1>;

describe("RouterView", () => {
  it("renders without crashing", () => {
    const routes = {
      "/": HomePage
    };

    const wrapper = mount(<RouterView routes={routes} />);

    expect(wrapper).toMatchSnapshot();
  });

  describe("renders 404 page", () => {
    it("with `*` route", () => {
      const routes = {
        "*": NotFoundPage
      };

      const wrapper = mount(
        <RouterView routes={routes} pathname="/not-found" />
      );

      expect(wrapper).toMatchSnapshot();
    });

    it("without `*` route", () => {
      const routes = {
        "/": HomePage
      };

      const wrapper = mount(
        <RouterView routes={routes} pathname="/not-found" />
      );

      expect(wrapper).toMatchSnapshot();
    });
  });

  it("should handleNav change pathname", () => {
    const routes = {
      "/": HomePage
    };

    const wrapper = mount(<RouterView routes={routes} pathname="/blog" />);

    expect(wrapper.state().pathname).toEqual("/blog");
    wrapper.instance().handleNav("/test");
    expect(wrapper.state().pathname).toEqual("/test");
  });
});
