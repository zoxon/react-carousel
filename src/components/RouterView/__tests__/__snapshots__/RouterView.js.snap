// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RouterView renders 404 page with \`*\` route 1`] = `
<RouterView
  basename="/"
  pathname="/not-found"
  routes={
    Object {
      "*": [Function],
    }
  }
>
  <NotFoundPage
    context={
      Object {
        "pathname": "/not-found",
      }
    }
  >
    <h1>
      Not found page
    </h1>
  </NotFoundPage>
</RouterView>
`;

exports[`RouterView renders 404 page without \`*\` route 1`] = `
<RouterView
  basename="/"
  pathname="/not-found"
  routes={
    Object {
      "/": [Function],
    }
  }
>
  <Component
    context={
      Object {
        "pathname": "/not-found",
      }
    }
  >
    <span>
      Page not found
    </span>
  </Component>
</RouterView>
`;

exports[`RouterView renders without crashing 1`] = `
<RouterView
  basename="/"
  pathname="/"
  routes={
    Object {
      "/": [Function],
    }
  }
>
  <HomePage
    context={
      Object {
        "pathname": "/",
      }
    }
  >
    <h1>
      Home page
    </h1>
  </HomePage>
</RouterView>
`;
